### PR TITLE
Feature/#128 create provider list

### PR DESCRIPTION
**Pull request by brauliodiez**
Added a subscriptionManager and a TrainingList Provider, advantages of this approach:

 - Separation of concerns (the ui component does not need to register to componentWillMount event thus can be a stateless component).
 - Smaller component.
 - In some cases we could have common providers.
 - This providers can be moved from a single Http Request, to pollling or event sockets if needed and is something transparent at application level.

I had some issues trying to make optional parameters, tried something like, but no luck when transpiling:

```javascript
  public static propTypes = {
    children: PropTypes.node.isRequired,
    payload: PropTypes.any,
    subscribe: PropTypes.func,
    unsubscribe: PropTypes.func,
  };


  public static defaultProps: Partial<Props>  = {
    payload: {},
    unsubscribe: () => {},
  };
```

Next Steps:

  - Propagate this  provider approach (including unit tests) to the rest of list like results and sing field like pages.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,7 +32,7 @@
     "@types/deep-freeze": "0.0.29",
     "@types/enzyme": "^2.5.38",
     "@types/history": "^3.2.1",
-    "@types/karma-chai-sinon": "^0.1.5",
+    "@types/karma-chai-sinon": "^0.1.5",
```
Comment (Nasdan) [inline package.json]: remove space

File: package.json [hunk 2]
```diff
@@ -89,7 +89,7 @@
   "dependencies": {
     "bootstrap": "^3.3.7",
     "jquery": "^3.1.1",
-    "lc-form-validation": "^1.0.0",
+    "lc-form-validation": "^1.0.0",
```
Comment (Nasdan) [inline package.json]: remove space

File: src/common/components/index.ts [hunk 3]
```diff
@@ -2,3 +2,4 @@ export { MarkDownViewerComponent } from './markdownViewer';
 export { PanelComponent, PanelItem } from './panel';
 export { Breadcrumb, BreadcrumbItem } from './breadcrumb';
 export { CounterButton } from './counterButton/counterButton';
+export { SubscriptionManager } from './subscriptionManager/subscriptionManager';
```
Comment (Nasdan) [inline src/common/components/index.ts]: create an index file to get something like `from './subscriptionManager'`

File: src/common/components/subscriptionManager/subscriptionManager.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  it('calls the subscribe action on componentWillMount', sinon.test(function() {
+    // Arrange
+    const sinon: sinon.SinonStatic = this;
+    const subscribeAction = sinon.spy((payload) => { });
+    const payload = {};
+    const unsubscribeAction = sinon.spy(() => { });
+
+    // Act
+    const progressBarComponent = mount(
```
Comment (Nasdan) [inline src/common/components/subscriptionManager/subscriptionManager.spec.tsx]: rename to `component`?

File: src/common/components/subscriptionManager/subscriptionManager.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  it('calls the unsubscribe action on componentUnmount', sinon.test(function() {
+    // Arrange
+    const sinon: sinon.SinonStatic = this;
+    const subscribeAction = sinon.spy((payload) => { });
+    const payload = {};
+    const unsubscribeAction = sinon.spy(() => { });
+
+    // Act
+    const progressBarComponent = mount(
```
Comment (Nasdan) [inline src/common/components/subscriptionManager/subscriptionManager.spec.tsx]: Same as above

File: src/common/components/subscriptionManager/subscriptionManager.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  props.subscribe(props.payload);
+};
+
+const unsubscribe = (props: Props) => {
+  if (props.unsubscribe) {
+    props.unsubscribe(props.payload);
+  }
+};
+
+interface Props  extends React.Props<Props> {
```
Comment (Nasdan) [inline src/common/components/subscriptionManager/subscriptionManager.tsx]: Why do you need to extend from React.Props?

File: src/common/components/subscriptionManager/subscriptionManager.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        subscribe={subscribeAction}
+        payload={payload}
+        unsubscribe={unsubscribeAction}
+       />,
+    );
+
+    progressBarComponent.unmount();
+
+    // Assert
+    expect(subscribeAction.called).to.be.true;
```
Comment (Nasdan) [inline src/common/components/subscriptionManager/subscriptionManager.spec.tsx]: Add expect to `calledWith` payload?

File: src/pages/trainer/training/list/page.tsx [hunk 8]
```diff
@@ -12,17 +13,19 @@ interface Props {
 
 export class TrainingListPage extends React.Component<Props, {}> {
   public componentDidMount() {
-    this.props.fetchTrainingList(this.props.trainerId);
+   // this.props.fetchTrainingList(this.props.trainerId);
```
Comment (Nasdan) [inline src/pages/trainer/training/list/page.tsx]: Remove comment

File: src/pages/trainer/training/list/spec/trainingListProvider.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import configureStore from 'redux-mock-store';
+import { IAppState } from '../../../../../reducers/index';
+import { TrainingState } from '../../../../../reducers/trainer/training';
+import { LoginState } from '../../../../../reducers/login';
+import * as actions from '../actions/trainingActions';
+import { TrainingSummary } from '../../../../../model/trainingSummary';
+import { TrainingListProvider } from '../trainingListProvider';
+import { SubscriptionManager } from '../../../../../common/components';
+
+const createStore = configureStore<IAppState>();
```
Comment (Nasdan) [inline src/pages/trainer/training/list/spec/trainingListProvider.spec.tsx]: What do you win when you type `configureStore<IAppState>`?

File: src/common/components/subscriptionManager/subscriptionManager.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import * as React from 'react';
+import { PropTypes } from 'react';
```
Comment (Nasdan) [inline src/common/components/subscriptionManager/subscriptionManager.tsx]: remove import

File: src/common/components/index.ts (hunk 3 above)
Comment (brauliodiez) [inline src/common/components/index.ts]: only used here, not sure if would be worth (is just one file and not expected to grow and only reference here in index)

File: src/common/components/subscriptionManager/subscriptionManager.tsx (hunk 10 above)
Comment (brauliodiez) [inline src/common/components/subscriptionManager/subscriptionManager.tsx]: done

File: src/pages/trainer/training/list/spec/trainingListProvider.spec.tsx (hunk 9 above)
Comment (brauliodiez) [inline src/pages/trainer/training/list/spec/trainingListProvider.spec.tsx]: Not sure, maybe nothing, will have to check later on

**Pull request merged**